text/0014-is-authorized-batch.md

```diff
@@ -0,0 +1,247 @@
+higher-level operation may involve multiple lower-level actions on multiple
+resources.
+
+In naive comparison testing with the
+[k9securityio/cedar-py](https://github.com/k9securityio/cedar-py)
+`is_authorized` implementation against a
+[very naive `is_authorized_batch` implementation](https://github.com/glb/cedar-py/commit/98bf3e1fa2e90bb6f7326cbd846a3293fdbd960b),
+I observed speedups of at least 5x for even simple cases, as documented in
+[k9securityio/cedar-py#13](https://github.com/k9securityio/cedar-py/issues/13).
+
+@skuenzli also
+[reported](https://github.com/k9securityio/cedar-py/issues/13#issuecomment-1629790294)
+that:
+
+> The time observed in Python is consistently 2 orders of magnitude greater than
+> the authorizer.is_authorized call. These particular requests use a simple
+> policy, a simple schema, and ~170 entities.
+
+The initial thought was to simply create an `is_authorized_batch` API for the
+`cedar-py` Python binding, but in discussion in the Cedar Policy Language Slack
+there was support from @aaronjeline and @andrewmwells-amazon for exposing this
+API in Cedar core so that it could be picked up more easily by other language
+bindings, as they expect those bindings to have similar needs and performance
+concerns.
+
+## Detailed design
+
+### Accept a batch of requests with the current shape
+
+This approach creates a new `is_authorized_batch` function, similar to the
+`is_authorized` function, that accepts a list of `Request` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L2006))
+and returns a list of `Response` structs
+([ref](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy/src/api.rs#L338)),
+which might have an implementation something like this:
+
+```rust
+impl Authorizer {
+    /// Returns a vector of authorization responses for the input vector of requests
+    /// with respect to the given `PolicySet` and `Entities`.
+    ///
+    /// The order of responses matches the order of requests.
+    pub fn is_authorized_batch(
+        &self,
+        requests: &Vec<Request>,
+        policy_set: &PolicySet,
+        entities: &Entities,
+    ) -> Vec<Response> {
+        let mut responses: Vec<Response> = vec![];
+
+        for request in requests {
+            let response = self.is_authorized(request, policy_set, entities);
+            responses.push(response);
+        }
+
+        responses
+    }
+}
+```
+
+In the special case where the input `requests` list is empty, the function
+should be a no-op and return an empty list.
+
+An advantage of this alternative is that it gives full flexibility to the API
+consumer in how they structure and batch requests.
+
+A disadvantage of this alternative is that for the initial motivating case of a
+`list` operation on a set of resources the principal, action, and context will
```

Comment:
What is the rational of having `context` part of the request instead of having a single `context` in batch_is_authorized?

Reply:
@Thor-Bjorgvinsson great question; I was trying to accommodate the general use case and didn't gather enough information about the motivation. My original proposal was the one shown in the alternatives where everything except the resources was static across all requests, but initial feedback was that it would be better to support a full request, and accepting a `Vec<Request>` was right there and relatively straightforward as a parallel construction to `is_authorized`'s signature. I should have asked more questions about what needed to vary to support their use case.

I expect that `principal` and `context` are likely to be constant across the set of requests in any given batch, but I don't have data or adequate input. I'll try to get some.

Reply:
From [this comment](https://github.com/cedar-policy/rfcs/pull/14#discussion_r1261827416) it looks like `context` could vary across the request set. I like the idea that an authorization request is the same shape whether it is being used in a single `is_authorized` call or it's being sent in a batch to `is_authorized_batch`.